packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -0,0 +1,89 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import ts from 'typescript';
+import * as util from '../util';
+
+type ExpressionWithTest =
+  | TSESTree.ConditionalExpression
+  | TSESTree.DoWhileStatement
+  | TSESTree.ForStatement
+  | TSESTree.IfStatement
+  | TSESTree.WhileStatement;
+
+export default util.createRule({
+  name: 'strict-boolean-expressions',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Restricts the types allowed in boolean expressions',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    schema: [],
+    messages: {
+      strictBooleanExpression: 'Unexpected non-boolean in conditional.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const service = util.getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    /**
+     * Determines if the node has a boolean type. Does recursion for nodes with
+     * left/right operators.
+     */
+    function isBooleanType(node: TSESTree.Expression): boolean {
+      if (node.type === AST_NODE_TYPES.LogicalExpression) {
+        return isBooleanType(node.left) && isBooleanType(node.right);
+      }
+
+      const tsNode = service.esTreeNodeToTSNodeMap.get<ts.ExpressionStatement>(
+        node,
+      );
+      const type = checker.getTypeAtLocation(tsNode);
```

Comment:
you should probably handle generics to some degree. This utility should do it for you.

Don't forget to add tests i.e.
```ts
function foo<T extends boolean>(arg: T) { return !arg; }
```

```suggestion
      const type = util.getConstrainedTypeAtLocation(checker, tsNode);
```